upgrade/upgrades: TestUpgradeSchemaChangerElements failed

upgrade/upgrades.TestUpgradeSchemaChangerElements [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8784960?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8784960?buildTab=artifacts#/) on master @ [fb1b2eceed332100b45728dcc3db0ade1c4e5544](https://github.com/cockroachdb/cockroach/commits/fb1b2eceed332100b45728dcc3db0ade1c4e5544):

Fatal error:

```
panic: test timed out after 14m55s

```
Stack: 

```
goroutine 361262 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2036 +0x88
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestUpgradeSchemaChangerElements
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/3538dab3e8bcdb43d30e6320ed019d4d/logTestUpgradeSchemaChangerElements2796237364
    test_log_scope.go:79: use -show-logs to present logs inline
=== RUN   TestUpgradeSchemaChangerElements/running_schema_change_that_will_have_a_deprecated_waitForSchemaChangerElementMigration_element
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-schema
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestUpgradeSchemaChangerElements.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-24697

upgrade/upgrades.TestUpgradeSchemaChangerElements [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8834788?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8834788?buildTab=artifacts#/) on master @ [647bd0b2ddfda9d8bf6488bc170ee2fafa1b5d95](https://github.com/cockroachdb/cockroach/commits/647bd0b2ddfda9d8bf6488bc170ee2fafa1b5d95):

Fatal error:

```
panic: test timed out after 14m55s

```
Stack: 

```
goroutine 309139 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2036 +0x88
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestUpgradeSchemaChangerElements
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/9a997cd2809ebcb0e38b3d249dc7df91/logTestUpgradeSchemaChangerElements2741179066
    test_log_scope.go:79: use -show-logs to present logs inline
=== RUN   TestUpgradeSchemaChangerElements/running_schema_change_that_will_have_a_deprecated_waitForSchemaChangerElementMigration_element
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestUpgradeSchemaChangerElements.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

upgrade/upgrades.TestUpgradeSchemaChangerElements [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8865697?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8865697?buildTab=artifacts#/) on master @ [528cd197bf43793672d0660dcabdd89676c0f753](https://github.com/cockroachdb/cockroach/commits/528cd197bf43793672d0660dcabdd89676c0f753):

Fatal error:

```
panic: test timed out after 14m55s

```
Stack: 

```
goroutine 304450 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2036 +0x88
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestUpgradeSchemaChangerElements
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/9a997cd2809ebcb0e38b3d249dc7df91/logTestUpgradeSchemaChangerElements2367157793
    test_log_scope.go:79: use -show-logs to present logs inline
=== RUN   TestUpgradeSchemaChangerElements/running_schema_change_that_will_have_a_deprecated_waitForSchemaChangerElementMigration_element
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestUpgradeSchemaChangerElements.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/upgrade/upgrades/schemachanger_elements_test.go
+++ b/pkg/upgrade/upgrades/schemachanger_elements_test.go
@@ -50,17 +50,16 @@ func TestUpgradeSchemaChangerElements(t *testing.T) {
 		waitType   int
 	}{
 		{
-			"running schema change that will have a deprecated waitForSchemaChangerElementMigration element",
-			nil,
-			func(t *testing.T, sqlDB *gosql.DB) error {
+			name: "running schema change that will have a deprecated waitForSchemaChangerElementMigration element",
+			run: func(t *testing.T, sqlDB *gosql.DB) error {
 				_, err := sqlDB.Exec("SET sql_safe_updates=off")
 				require.NoError(t, err)
 				_, err = sqlDB.Exec("ALTER TABLE t DROP COLUMN b")
 				require.ErrorContains(t, err, "was paused before it completed with reason: pause point \"newschemachanger.before.exec\" hit")
 				return nil
 			},
-			true,
-			WaitForSchemaChangeCompletion,
+			unpauseJob: true,
+			waitType:   WaitForSchemaChangeCompletion,
 		},
 		{
 			"running schema change that will have no deprecated elements",
@@ -117,7 +116,7 @@ func TestUpgradeSchemaChangerElements(t *testing.T) {
 				schemaChangeAllowedToComplete chan struct{}
 				waitedForJob                  chan struct{}
 				jobIsPaused                   chan struct{}
-				readyToQuery                  int64
+				readyToQuery                  int64 = 0
 			)
 			scJobID := int64(jobspb.InvalidJobID)
 			params.Knobs.SQLDeclarativeSchemaChanger = &scexec.TestingKnobs{
@@ -142,9 +141,8 @@ func TestUpgradeSchemaChangerElements(t *testing.T) {
 				if targetJobID := jobspb.JobID(atomic.LoadInt64(&scJobID)); targetJobID != jobspb.InvalidJobID {
 					for _, j := range jobs {
 						// The upgrade is waiting for the job...
-						if j == targetJobID && atomic.LoadInt64(&readyToQuery) > 0 {
+						if j == targetJobID && atomic.CompareAndSwapInt64(&readyToQuery, 1, 0) {
 							atomic.StoreInt64(&scJobID, int64(jobspb.InvalidJobID))
-							atomic.StoreInt64(&readyToQuery, 0)
 							waitedForJob <- struct{}{}
 						}
 					}
@@ -167,7 +165,6 @@ func TestUpgradeSchemaChangerElements(t *testing.T) {
 			schemaChangeAllowedToComplete = make(chan struct{})
 			waitedForJob = make(chan struct{})
 			jobIsPaused = make(chan struct{})
-			atomic.StoreInt64(&readyToQuery, 0)
 			g.GoCtx(func(ctx context.Context) error {
 				err := tc.run(t, sqlDB)
 				jobIsPaused <- struct{}{}

